Pond model

kmdeck opened the pull request:
Add a pond model. this mimicks pieces of both the river and snow model in terms of domain (surface only) and how it will eventually interact with the soil model.

kmdeck commented on src/Pond.jl:
```diff
@@ -0,0 +1,119 @@
+    println(typeof(i_c))
+    println(typeof(Y.pond.η))
+    println(typeof(boundary_exchanges.P(t)))
+    infiltration =
```
@jb-mackay This was an example where I had to use `Y.pond.eta[1]` a lot originally. 
```
 i_c =
        hydraulic_conductivity(Ksat, m, S) *
        (max(FT(0.0), Y.pond.η[1]) - matric_potential(α, n, m, S) + Δz) / Δz
    if Y.pond.η[1] > eps(FT) # when there is a pond, infiltration is pond/soil limited
        infiltration = i_c
    else # when there is no pond, infiltration is precip/soil limited
        infiltration = min(i_c, boundary_exchanges.P(t))
    end
```

But I tried to rewrite it so we do not...it seems like it should work, but this line is leading to errors - only when run within the function.  calling `infiltration_at_point` by itself with objects of the same type `infiltration_at_point.(a::Vector{FT}, b::Vector{FT}, c::FT)` does not produce an error.

kmdeck replied:
fixed this! all set

jb-mackay commented on src/Domains.jl:
```diff
@@ -48,9 +47,19 @@ function coordinates(domain::RootDomain{FT}) where {FT}
+surface of the soil domain.
+"""
+struct SurfaceDomain{FT} <: AbstractDomain{FT}
+    height::Vector{FT}
```
Can you explain this a bit and its use cases? I'm a little confused that a 2D surface domain has a `height` field.

kmdeck replied:
hm, so surface variables are defined on the 2d surface `(x,y, z = z_sfc(x,y))`, where `height = z_sfc` (I can rename `height` if we keep this).  There will be topography eventually. 

I think there are two questions here.
1. Does the model require this `z` info to compute RHSs? For example,  the soil model needs topography, `z_sfc(x,y)`, and should have a warped surface to the domain, but Im not sure the actual surface models need it because I dont think `z` shows up in the RHS terms for them. *As long as the correct surface variables are linked up to the right column of soil (in terms of horizontal location - I dont know how we will do this in 3D LSMs), we may not need the surface coordinates at all in compute RHS's.*

2. We initialize variables by computing `coordinates(domain)` and (`similar(coords)`). So we need something here...whether it is `(x,y)` or `(x,y, z_sfc(x,y))`. But right now we are only working with columns, so we only have a notion of `z`?

kmdeck replied:
for the first question - we can use a Plane Domain, and give it a single element and single nodal point -> single set of (x,y) is returned by coordinates(model)

The second question is still pertinent! It makes sense to return coordinates as (x,y) pair, but then we cant do similar(coordinates), because coordinates acts like a NamedTuple. this also came up with the 3d domain for soil - returning (x,y,z) coordinates, but wanting similar(coords.z) e.g.

kmdeck replied:
fixed by doing
zero_state = map(_ -> zero(FT), coords)
and then do similar(zero_state), instead of similar(coords)

kmdeck commented:
branch is WIP - rebased to new main, simplified pond model for our example, rebuilt based on changes we've made. Got overzealous and added in 2d spectral element domain for the pond, whcih we can make have a single nodal point and single element for "column" mode. Try to get this to run with pond standalone, then go back to pond_soil_lsm...

kmdeck commented:
Once we merge this: https://github.com/CliMA/ClimaLSM.jl/pull/22
this PR will get a lot smaller (no domain changes, coordinate changes)
and then we can discuss whether or not the pond domain should be a plane or something else...

we could also try and add in the pond_soil_lsm model as well...

jb-mackay commented on src/SurfaceWater/Pond.jl:
```diff
@@ -0,0 +1,67 @@
+as a function of time, and soil parameters, which affect infiltration.
+"""
+struct PrescribedRunoff{FT} <: AbstractSurfaceRunoff{FT}
+    "Time dependent P⃗ = Pẑ, given in m/s"
```
"Time dependent precipitation P = ..."

jb-mackay commented on src/pond_soil_model.jl:
```diff
@@ -0,0 +1,117 @@
+end
+
+struct PrognosticRunoff{FT} <: Pond.AbstractSurfaceRunoff{FT}
+    precip::Function
```
This is called `P` in the PrescrivedRunoff case (unify)

jb-mackay replied:
use `precip` as there are other `p`s

jb-mackay commented on test/pond_soil_lsm.jl:
```diff
@@ -0,0 +1,56 @@
+    Ysoil.soil.ϑ_l .= hydrostatic_profile.(z, Ref(params))
+end
+init_soil!(Y, coords.soil, land.soil.param_set)
+Y.surface_water.η .= 0.0
```
`init_pond`

kmdeck replied:
added comment indicating this.

jb-mackay commented on README.md:
```diff
@@ -1,5 +1,38 @@
+Examples:
+
+Component Models:
+RichardsModel <: AbstractModel [runnable w/o LandModel wrapper as well]
```
RichardsModel <: AbstractSoilModel <: AbstractModel

kmdeck commented on test/pond_soil_lsm.jl:
```diff
@@ -0,0 +1,56 @@
+dt = FT(1);
+cb = SavingCallback((u, t, integrator) -> integrator.p, saved_values)
+prob = ODEProblem(ode!, Y, (t0, tf), p);
+sol = solve(prob, Euler(), dt = dt, callback = cb);
```
this is fine, but we should test some of our infiltration functions too & make sure they work OK in all cases!
